### PR TITLE
Clean up bundle config generation for Samza Runner

### PR DESCRIPTION
**Pull request by xinyuiscool**
Refactor the samza runner configs into a new function with consideration of autosizing configs. Also add unit test for it.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by xinyuiscool** [comment]
R: @mynameborat please review. Thanks.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java
Comment (xinyuiscool) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:132]: Added.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java
Comment (xinyuiscool) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: If the user sets the pipeline options in code, it should take higher precedence. Otherwise it will be quite confusing. Autosizing will work if the user doesn't set pipeline options.

We set threadPoolSize to 0 regardless.

File: runners/samza/src/test/java/org/apache/beam/runners/samza/translation/ConfigGeneratorTest.java
Comment (xinyuiscool) [reply runners/samza/src/test/java/org/apache/beam/runners/samza/translation/ConfigGeneratorTest.java:422]: Good point. I added the case in the same tests. The results are expected to be the same.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java
Comment (mynameborat) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: Isn't this opposite to the existing pattern we have for autosizing? i.e., users can configure resourcing but if they enable autosizing, we treat autosizing overrides as the final values.

This would be inconsistent with the existing behavior for every other auto-sizer tunable configurations.

Comment (xinyuiscool) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: Yes, autosizing will override the same user config. But this is a pipeline option in code. It'll be very confusing if we set some parameters in code while it won't be taken into effect in the runtime. To me, this is definitely higher in precedence. The other problem is that we need a way for user to override without turning off autosizing. Our autosizing hasn't been tuned for thread pool size with bundle > 1. So we need to have this capability around.

Comment (mynameborat) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: I feel the inconsistency in behavior is as well confusing to the users. If we want to establish user code as the highest precedence, maybe we should not tie autosizing `job.autosizing.container.thread.pool.size` into this.

Consider this example
1. user configures `job.container.thread.pool.size`, it is reflected in the application behavior
2. autosizing is enabled and user configuration no longer is reflected in the application behavior
3. User sets it in code and the change is reflected in application behavior

user configuration << framework configuration override << user code 

Also consider autosizing devs/support team not realizing knobs x, y and z follows auto-sizer values as higher precedence on their end vs a, b follows user values as higher precedence.


>The other problem is that we need a way for user to override without turning off autosizing
I think the recommended way for now is to use auto-sizer tool to overrides specific knobs.

Comment (jia-gao) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: To clarify, ` job.container.thread.pool.size` won't work if auto-sizing is enabled.
auto sizing today allows user overrides size configs but still leave auto sizing enabled (asc can still resize the job)  by:
1. users can configure `job.autosizing.container.thread.pool.size`. It will be honored when the job starts
2. when auto-sizing resizes the job, it will submit the job to jetty proxy with a new value of 
    `job.autosizing.container.thread.pool.size` included in the submission config. and write it to the coordinator 
   stream
3. If user redeploys the job again, the size in 1 is still honored. If the user removes the config in 1 and redeploy, 
   the  size in 2 will be honored (reading from coordinator stream)

for beam job options like ThreadsForProcessElement, asc is not able to achieve 2 so I agree that in ConfigBuilder we should honor the `job.autosizing.container.thread.pool.size` over options specified in code to be able to resize the job. Afterall, if `job.autosizing.container.thread.pool.size` is present, if means either asc resizes the job or the job owner wants to override the size but still want to leave auto-sizing on

Comment (jia-gao) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: > I think the recommended way for now is to use auto-sizer tool to overrides specific knobs.
asc-tool does the same thing as step 2 above. ConfigBuilder still needs to address the problem job config vs pipeline option precedence here

Comment (xinyuiscool) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: Synced up offline with @mynameborat  and @jia-gao . We will go with this option precedence approach first to unblock jobs onboarding and then follow up with samza autosizing to figure out precedence.

Comment (mynameborat) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java:140]: We sync'd up offline on this. The viewpoint to have autosizing enabled plus allow user overrides for knobs that autosizing hasn't caught up on is very valid scenario.

We don't have such compartmentalization today as there are relationships between policies dealing with different resources and hence disabling autosizing all out is the current recommendation to override. That said, unblocking this PR for site-up and we can follow up on this

**Review by mynameborat — APPROVED** [review]

**Comment by mynameborat** [comment]
@xinyuiscool Can you look into the test failure before you merge?

**Comment by xinyuiscool** [comment]
> @xinyuiscool Can you look into the test failure before you merge?

Yup, seems some env issues in the windows OS test. Will wait till Java PreCommit pass.

**Comment by xinyuiscool** [comment]
Run Java PreCommit

**Pull request merged**
